0006-ffmul-revised.md

```diff
@@ -0,0 +1,518 @@
+
+* **Introduce 1 witness**: $q'_2$
+
+The addition is another constraint:
+
+$$\tag{C11: bound for $q$}
+q'_2 = q_2 + (2^\ell - f_2 - 1) \mod n
+$$
+
+And a range check on $q'_2$ completes the bounds check on $q$:
+
+$$\tag{RC: bound for $q$}
+q'_2 \in [0,2^\ell)
+$$
+
+This implies that $a,b,q \in[0, 2^{2\ell}(f_2 + 1))$. For $r$, we already know that $r \in [0, 2^{3\ell})$.
+
+We now get the following upper bound and lower bound:
+
+$$
+ab - qf - r \le ab < 2^{4\ell}(f_2 + 1)^2
+$$
+
+$$
+-ab + qf + r \le qf + r < 2^{4\ell}(f_2 + 1)^2 + 2^{3\ell}
+$$
+
+In summary, we have
+
+$$
+|ab - qf - r| < 2^{4\ell}(f_2 + 1)^2 + 2^{3\ell} = 2^{3\ell}\cdot(2^\ell (f_2 + 1)^2 + 1)
+$$
+
+In the case that $f < 2^{259}$, we have $(f_2 + 1) \le 2^{259 - 2\ell} = 2^{83}$, and our estimate works out to be 
+
+$$
+|ab - qf - r| < 2^{3\ell} \cdot (2^{254} + 1)
+$$
+
+For $n > 2^{254}$, wich is true for both Pasta moduli, we conclude that
+
+$$
+|ab - qf - r| < 2^{3\ell}n
+$$
+
+and so $ab = qf + r$, which proves the soundness of the ffmul gate, when used together with the listed external checks.
+
+### Gate layout
+
+The following 14 witnesses have to be made available to other gates:
+* The 6 input limbs $a_0, a_1, a_2$ and $b_0, b_1, b_2$
+* The 5 output limbs $q_0, q_1, q_2$ and $r_{01}$, $r_2$
+* Intermediate values $p_{10}$ and $p_{110}$ for the RC
+* The high-limb qotient bound $q'_2$ for the RC
+
+To make them available, they have to go in permutable cells. Kimchi has 7 permutable cells per row, so we exactly fit these values in the current and next row of a single-row gate.
+
+
+| FFmul  | 0p | 1p | 2p | 3p | 4p | 5p | 6p | 7  | 8  | 9  | 10 | 11 | 12 | 13 | 14 |
+| -------| -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- |
+| _Curr_ | $a_0$ | $a_1$ | $a_2$ | $b_0$ | $b_1$ | $b_2$ | $p_{10}$ |  |  |  |  |  |  |  |  | 
+| _Next_ | $r_{01}$ | $r_2$ | $q_0$ | $q_1$ | $q_2$ | $q'_2$ | $p_{110}$ |  |  |  |  |  |  |  | 
+
+Another constraint on the gate layout is that we can only do 4 lookups per row, and we need 7 lookups on the carry chunks $c_{1,0},\ldots,c_{1,72}$. So these have to be divided up between the two rows.
+
+The remaining witnesses are just put into any free cells to define the final gate layout:
+
+| FFmul  | 0p | 1p | 2p | 3p | 4p | 5p | 6p | 7  | 8  | 9  | 10 | 11 | 12 | 13 | 14 |
+| -------| -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- |
+| _Curr_ | $a_0$ | $a_1$ | $a_2$ | $b_0$ | $b_1$ | $b_2$ | $p_{10}$ | $c_{1,0}$ | $c_{1,12}$ | $c_{1,24}$ | $c_{1,36}$ | $c_{1,84}$ | $c_{1,86}$ | $c_{1,88}$ | $c_{1,90}$ | 
+| _Next_ | $r_{01}$ | $r_2$ | $q_0$ | $q_1$ | $q_2$ | $q'_2$ | $p_{110}$ | $c_{1,48}$ | $c_{1,60}$ | $c_{1,72}$ | $c_0$ |  |  |  | 
+
+The limbs of $f' = 2^{3\ell} - f$, which feature in a few of the constraints, are embedded into the gate as coefficients.
```

Comment:
Some constraints (e.g. `C1`) refer to `f` rather than `f'`. Can you update the constraints to use `f'` if we won't be storing `f`, or otherwise propose storing the limbs of `f`?

Reply:
Good point!